thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/QuestionService.java

```diff
@@ -1,9 +1,19 @@
 package pl.edu.agh.quizzesthesis.business.service;
 
+import lombok.AllArgsConstructor;
 import org.springframework.transaction.annotation.Transactional;
+import pl.edu.agh.quizzesthesis.business.UserAuthDetails;
 
-public interface QuestionService<T> {
+@AllArgsConstructor
+public abstract class QuestionService<T> {
+
+    private final UserService userService;
 
     @Transactional
-    T generateQuestion(int categoryId, Integer difficulty);
+    public abstract T generateQuestion(int categoryId, UserAuthDetails userAuthDetails);
+
+    protected float getUserRankInCategory(UserAuthDetails userAuthDetails, int categoryId) {
```

Comment:
Honestly, I don't know what to think about this. Right now our `QuestionService` is so big (it requires 5 different services), but I don't know if we can do something about this. I especially don't like including `UserService` in `QustionService` (`QuestionService` should not care who it is creating question for, it only needs information about the difficulty of the term... take it back, it only needs information  for what term it needs to generate question). One way for fixing that is to move this method to `UserService` and invoke it in the `QuestionController`, passing the returned value to `QuestionService`. Cons - it require `QuestionController` to have another service (that is not directly connected to questions) and will have to invoke its method, which means that controller is responsible for some inside logic. Another idea is to move this to `TermService` (so that `TermSerive` could return proper term to `QuestionService`). Cons - we are basically transferring responsibility from one service to another.

Reply:
Yeah, I understand your concern. QuestionService has many dependencies, but it isn't huge itself. Its one method `generateQuestion` does not contain much logic beyond calling other services to produce a question. I would not move calling UserService to QuestionController - if QuestionService shouldn't know about users, then QuestionController shouldn't know either. Moving it to TermService also doesn't solve anything - why should TermService know about the fact that term difficulty should be equal to user rank? The call to UserService has to be somewhere and if there is no suitable place, I think the best idea is to put it in the highest-level service - QuestionService.

Reply:
After second thought and reading your response I can agree with your decision. It might not be the nicest solution, but it is probably the best considering our current situation.